File: GpuData/Core+PCache1/12.PCache1+General.cs
using System;
using Pgno = System.UInt32;
using System.Diagnostics;
using System.Text;
namespace Core
{
    public partial class PCache1
    {
        private RC pcache1ResizeHash()
        {
            Debug.Assert(MutexEx.Held(Group.Mutex));
            var nNew = nHash * 2;
            if (nNew < 256)
                nNew = 256;
            LeaveMutex(Group);
            if (nHash != 0)
                MallocEx.BeginBenignMalloc();
            var apNew = new PgHdr1[nNew];
            if (nHash != 0)
                MallocEx.EndBenignMalloc();
            EnterMutex(Group);
            if (apNew != null)
            {
                for (var i = 0; i < nHash; i++)
                {
                    var pNext = Hash[i];
                    PgHdr1 pPage;
                    while ((pPage = pNext) != null)
                    {
                        var h = (Pgno)(pPage.Key % nNew);
                        pNext = pPage.Next;
                        pPage.Next = apNew[h];
                        apNew[h] = pPage;
                    }
                }
                Hash = apNew;
                nHash = nNew;
            }
            return (Hash != null ? RC.OK : RC.NOMEM);
        }

        private static void pcache1PinPage(PgHdr1 pPage)
        {
            if (pPage == null)
                return;
            var pCache = pPage.Cache;
            var pGroup = pCache.Group;
            Debug.Assert(MutexEx.Held(pGroup.Mutex));
            if (pPage.LruNext != null || pPage == pGroup.LruTail)
            {
                if (pPage.LruPrev != null)
                    pPage.LruPrev.LruNext = pPage.LruNext;
                if (pPage.LruNext != null)
                    pPage.LruNext.LruPrev = pPage.LruPrev;
                if (pGroup.LruHead == pPage)
                    pGroup.LruHead = pPage.LruNext;
                if (pGroup.LruTail == pPage)
                    pGroup.LruTail = pPage.LruPrev;
                pPage.LruNext = null;
                pPage.LruPrev = null;
                pPage.Cache.Recyclables--;
            }
        }

        private static void pcache1RemoveFromHash(PgHdr1 pPage)
        {
            var pCache = pPage.Cache;
            Debug.Assert(MutexEx.Held(pCache.Group.Mutex));
            var h = (int)(pPage.Key % pCache.nHash);
            PgHdr1 pPrev = null;
            PgHdr1 pp;
            for (pp = pCache.Hash[h]; pp != pPage; pPrev = pp, pp = pp.Next) ;
            if (pPrev == null)
                pCache.Hash[h] = pp.Next;
            else
                pPrev.Next = pp.Next;
            pCache.Pages--;
        }

        private static void pcache1EnforceMaxPage(PGroup pGroup)
        {
            Debug.Assert(MutexEx.Held(pGroup.Mutex));
            while (pGroup.CurrentPages > pGroup.MaxPages && pGroup.LruTail != null)
            {
                PgHdr1 p = pGroup.LruTail;
                Debug.Assert(p.Cache.Group == pGroup);
                pcache1PinPage(p);
                pcache1RemoveFromHash(p);
                pcache1FreePage(ref p);
            }
        }

        private void pcache1TruncateUnsafe(uint iLimit)
        {
#if !DEBUG
            uint nPage = 0;
#endif
            Debug.Assert(MutexEx.Held(Group.Mutex));
            for (uint h = 0; h < nHash; h++)
            {
                var pp = Hash[h];
                PgHdr1 pPrev = null;
                PgHdr1 pPage;
                while ((pPage = pp) != null)
                {
                    if (pPage.Key >= iLimit)
                    {
                        Pages--;
                        pp = pPage.Next;
                        pcache1PinPage(pPage);
                        if (Hash[h] == pPage)
                            Hash[h] = pPage.Next;
                        else
                            pPrev.Next = pp;
                        pcache1FreePage(ref pPage);
                    }
                    else
                    {
                        pp = pPage.Next;
#if !DEBUG
                        nPage++;
#endif
                    }
                    pPrev = pPage;
                }
            }
#if !DEBUG
            Debug.Assert(pCache.nPage == nPage);
#endif
        }
    }
}

File: src/GpuData/Core+PCache1/12.PCache1+General.cpp
namespace Core
{
	/*
** This function is used to resize the hash table used by the cache passed
** as the first argument.
**
** The PCache mutex must be held when this function is called.
*/
static int pcache1ResizeHash(PCache1 *p){
  PgHdr1 **apNew;
  unsigned int nNew;
  unsigned int i;

  assert( sqlite3_mutex_held(p->pGroup->mutex) );

  nNew = p->nHash*2;
  if( nNew<256 ){
    nNew = 256;
  }

  pcache1LeaveMutex(p->pGroup);
  if( p->nHash ){ sqlite3BeginBenignMalloc(); }
  apNew = (PgHdr1 **)sqlite3MallocZero(sizeof(PgHdr1 *)*nNew);
  if( p->nHash ){ sqlite3EndBenignMalloc(); }
  pcache1EnterMutex(p->pGroup);
  if( apNew ){
    for(i=0; i<p->nHash; i++){
      PgHdr1 *pPage;
      PgHdr1 *pNext = p->apHash[i];
      while( (pPage = pNext)!=0 ){
        unsigned int h = pPage->iKey % nNew;
        pNext = pPage->pNext;
        pPage->pNext = apNew[h];
        apNew[h] = pPage;
      }
    }
    sqlite3_free(p->apHash);
    p->apHash = apNew;
    p->nHash = nNew;
  }

  return (p->apHash ? SQLITE_OK : SQLITE_NOMEM);
}

/*
** This function is used internally to remove the page pPage from the 
** PGroup LRU list, if is part of it. If pPage is not part of the PGroup
** LRU list, then this function is a no-op.
**
** The PGroup mutex must be held when this function is called.
**
** If pPage is NULL then this routine is a no-op.
*/
static void pcache1PinPage(PgHdr1 *pPage){
  PCache1 *pCache;
  PGroup *pGroup;

  if( pPage==0 ) return;
  pCache = pPage->pCache;
  pGroup = pCache->pGroup;
  assert( sqlite3_mutex_held(pGroup->mutex) );
  if( pPage->pLruNext || pPage==pGroup->pLruTail ){
    if( pPage->pLruPrev ){
      pPage->pLruPrev->pLruNext = pPage->pLruNext;
    }
    if( pPage->pLruNext ){
      pPage->pLruNext->pLruPrev = pPage->pLruPrev;
    }
    if( pGroup->pLruHead==pPage ){
      pGroup->pLruHead = pPage->pLruNext;
    }
    if( pGroup->pLruTail==pPage ){
      pGroup->pLruTail = pPage->pLruPrev;
    }
    pPage->pLruNext = 0;
    pPage->pLruPrev = 0;
    pPage->pCache->nRecyclable--;
  }
}


/*
** Remove the page supplied as an argument from the hash table 
** (PCache1.apHash structure) that it is currently stored in.
**
** The PGroup mutex must be held when this function is called.
*/
static void pcache1RemoveFromHash(PgHdr1 *pPage){
  unsigned int h;
  PCache1 *pCache = pPage->pCache;
  PgHdr1 **pp;

  assert( sqlite3_mutex_held(pCache->pGroup->mutex) );
  h = pPage->iKey % pCache->nHash;
  for(pp=&pCache->apHash[h]; (*pp)!=pPage; pp=&(*pp)->pNext);
  *pp = (*pp)->pNext;

  pCache->nPage--;
}

/*
** If there are currently more than nMaxPage pages allocated, try
** to recycle pages to reduce the number allocated to nMaxPage.
*/
static void pcache1EnforceMaxPage(PGroup *pGroup){
  assert( sqlite3_mutex_held(pGroup->mutex) );
  while( pGroup->nCurrentPage>pGroup->nMaxPage && pGroup->pLruTail ){
    PgHdr1 *p = pGroup->pLruTail;
    assert( p->pCache->pGroup==pGroup );
    pcache1PinPage(p);
    pcache1RemoveFromHash(p);
    pcache1FreePage(p);
  }
}

/*
** Discard all pages from cache pCache with a page number (key value) 
** greater than or equal to iLimit. Any pinned pages that meet this 
** criteria are unpinned before they are discarded.
**
** The PCache mutex must be held when this function is called.
*/
static void pcache1TruncateUnsafe(
  PCache1 *pCache,             /* The cache to truncate */
  unsigned int iLimit          /* Drop pages with this pgno or larger */
){
  TESTONLY( unsigned int nPage = 0; )  /* To assert pCache->nPage is correct */
  unsigned int h;
  assert( sqlite3_mutex_held(pCache->pGroup->mutex) );
  for(h=0; h<pCache->nHash; h++){
    PgHdr1 **pp = &pCache->apHash[h]; 
    PgHdr1 *pPage;
    while( (pPage = *pp)!=0 ){
      if( pPage->iKey>=iLimit ){
        pCache->nPage--;
        *pp = pPage->pNext;
        pcache1PinPage(pPage);
        pcache1FreePage(pPage);
      }else{
        pp = &pPage->pNext;
        TESTONLY( nPage++; )
      }
    }
  }
  assert( pCache->nPage==nPage );
}

}
